crates/nu-command/src/filters/chunk_by.rs

```diff
@@ -0,0 +1,220 @@
+    }
+
+    fn run(
+        &self,
+        engine_state: &EngineState,
+        stack: &mut Stack,
+        call: &Call,
+        input: PipelineData,
+    ) -> Result<PipelineData, ShellError> {
+        chunk_by(engine_state, stack, call, input)
+    }
+
+    fn examples(&self) -> Vec<Example> {
+        vec![
+            Example {
+                description: "Chunk data into runs of larger than zero or not.",
+                example: "[1, 3, -2, -2, 0, 1, 2] | chunk-by {|it| $it >= 0 }",
+                result: Some(Value::test_list(vec![
+                    Value::test_list(vec![Value::test_int(1), Value::test_int(3)]),
+                    Value::test_list(vec![Value::test_int(-2), Value::test_int(-2)]),
+                    Value::test_list(vec![
+                        Value::test_int(0),
+                        Value::test_int(1),
+                        Value::test_int(2),
+                    ]),
+                ])),
+            },
+            Example {
+                description: "Identify repetitions in a string",
+                example: r#"[a b b c c c] | chunk-by { |it| $it }"#,
+                result: Some(Value::test_list(vec![
+                    Value::test_list(vec![Value::test_string("a")]),
+                    Value::test_list(vec![Value::test_string("b"), Value::test_string("b")]),
+                    Value::test_list(vec![
+                        Value::test_string("c"),
+                        Value::test_string("c"),
+                        Value::test_string("c"),
+                    ]),
+                ])),
+            },
+        ]
+    }
+}
+
+struct Chunk<I, T, F, K> {
+    iterator: I,
+    last_value: Option<(T, K)>,
+    closure: F,
+    done: bool,
+}
+
+impl<I, T, F, K> Iterator for Chunk<I, T, F, K>
+where
+    I: Iterator<Item = T>,
+    F: FnMut(&T) -> K,
+    K: PartialEq,
+{
+    type Item = Vec<T>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.done {
+            return None;
+        }
+
+        let (head, head_key) = match self.last_value.take() {
+            None => {
+                let head = self.iterator.next()?;
+
+                let key = (self.closure)(&head);
+
+                (head, key)
+            }
+
+            Some((value, key)) => (value, key),
+        };
+
+        let mut result = vec![head];
+
+        loop {
+            match self.iterator.next() {
+                None => {
+                    self.done = true;
+                    return Some(result);
+                }
+                Some(value) => {
+                    let value_key = (self.closure)(&value);
+
+                    if value_key == head_key {
+                        result.push(value);
+                    } else {
+                        self.last_value = Some((value, value_key));
+                        return Some(result);
+                    }
+                }
+            }
+        }
+    }
+}
```

Comment:
The `Chunk` iterator should probably hold a copy of `EngineState::signals` and check it inside this loop to respond to interupts/ctrl+c.